app/src/main/java/com/example/drawingactivity/GradientView.kt

```diff
@@ -0,0 +1,166 @@
+                    onSatValChanged?.invoke(dx / width, 1 - (dy / height))
+                    invalidate()
+                }
+                MotionEvent.ACTION_UP -> {
+                    onGradientClicked?.invoke(false)
+                }
+            }
+            true
+        }
+    }
+
+    fun setHue(huePickerValue: Float) {
+        currentHue = huePickerValue
+        invalidate()
+    }
+
+    fun setSaturationAndValue(saturation: Float, value: Float) {
+        dx = saturation * width
+        dy = value * height
+        invalidate()
+    }
+
+    override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+        canvas?.apply {
+            if (shape == null) {
+                shape = Rect(0, 0, width, height)
+            }
+            generateAllBitmaps()
+            getBitmap(currentHue.toInt())?.let { bitmap ->
+                shape?.let { shape ->
+                    drawBitmap(
+                        bitmap,
+                        null,
+                        shape,
+                        null
+                    )
+                }
+            }
+            drawCircle(dx, height - dy, 12f, selectPaint)
+        }
+    }
+
+    companion object GradientConstants {
+        private const val imageSize: Int = 50
+        private const val maxPossibleHue: Int = 360
+        private val selectPaint = Paint().apply {
+            strokeWidth = 3f
+            color = BLACK
+            style = Paint.Style.STROKE
+        }
+        private var bitmaps: Array<Bitmap?>? = Array(maxPossibleHue) { null }
+        var stillCalculating = false
+
+        @OptIn(DelicateCoroutinesApi::class)
+        private fun generateAllBitmaps() {
+            if (!stillCalculating) {
+                GlobalScope.launch {
+                    withContext(Dispatchers.Default) {
+                        for (i in 0..90) {
+                            bitmaps?.let {bitmaps -> bitmaps[i] = makeBitmap(i.toFloat()) }
+                        }
+                    }
+                }
+                GlobalScope.launch {
+                    withContext(Dispatchers.Default) {
+                        for (i in 91..180) {
+                            bitmaps?.let {bitmaps -> bitmaps[i] = makeBitmap(i.toFloat()) }
+                        }
+                    }
+                }
+                GlobalScope.launch {
+                    withContext(Dispatchers.Default) {
+                        for (i in 181..270) {
+                          bitmaps?.let {bitmaps -> bitmaps[i] = makeBitmap(i.toFloat()) }
+                        }
+                    }
+                }
+                GlobalScope.launch {
+                    withContext(Dispatchers.Default) {
+                        for (i in 271..359) {
+                            bitmaps?.let {bitmaps -> bitmaps[i] = makeBitmap(i.toFloat()) }
+                        }
+                        stillCalculating = true
```

Comment:
Why's only this coroutine have this?

Reply:
moved it to a lower bracket so that once the process is started, it isn't recalled needlessly